README.md
chillout.js
===========

[**README (日本語)**](README-ja.md)

Reduce CPU usage by non-blocking asynchronous loop and psychologically speed up to improve the user experience in JavaScript.

[![NPM Version](https://img.shields.io/npm/v/chillout.svg)](https://www.npmjs.com/package/chillout)
[![Build Status](https://travis-ci.org/polygonplanet/chillout.svg?branch=master)](https://travis-ci.org/polygonplanet/chillout)
[![Bundle Size (minified)](https://img.shields.io/github/size/polygonplanet/chillout/dist/chillout.min.js.svg)](https://github.com/polygonplanet/chillout/blob/master/dist/chillout.min.js)
[![GitHub License](https://img.shields.io/github/license/polygonplanet/chillout.svg)](https://github.com/polygonplanet/chillout/blob/master/LICENSE)

## Table of contents

* [Overview](#overview)
* [Installation](#installation)
* [Compatibility](#compatibility)
* [async / await](#async--await)
* [Benchmarks](#benchmarks)
* [API](#api)
  + [forEach](#foreach)
  + [repeat](#repeat)
  + [until](#until)
  + [waitUntil](#waituntil)
  + [forOf](#forof)
* [Comparison Table](#comparison-table)
* [Contributing](#contributing)
* [License](#license)

## Overview

Unlike general accelerate way of the "to shorten the processing time physically", chillout.js speed up the JavaScript loops psychologically by reduce CPU usage and release resources to improve user experience.

### In order to improve slow processing

We feel stress when page load or processing is slow and the screen is lagging.
Moreover, there is a risk that the machine will down due to overheating because it can't keep up CPU cooling when continues running the heavy-load application with high CPU utilization rate.

### How to reduce CPU load in JavaScript?

Most of the slow processing is caused by looping that have deep nested looping.
If we think simply, it should wait for a little time like "sleep" in the looping, but we can't wait, because JavaScript have not "sleep" function. Then how to sleep in JavaScript? We can sleep by using `setTimeout` or `process.nextTick` with `Promise` in "asynchronous" processing.

### Speed up the JavaScript loops psychologically not physically

We can feel like "fast" or "interesting" psychologically by relieving user stress, even if the processing is slow if your application or game will used by humans not AI.

### About chillout.js

Provides asynchronous iteration functions that have a `Promise` based interface and it can run with low CPU usage.
In most cases, you can reduce the CPU usage by using the non-blocking loop of chillout.js API instead native JavaScript loops in your code, but you need to change it to asynchronous loops.

Chillout.js adds a little delay on each iteration if the processing is slow to maintain the CPU stability, and it continues iterate without delay if processing is fast.
Therefore, it will realize friendly processing for your machine.
And, it can execute JavaScript without **"Warning: Unresponsive Script"** alert in the browser.

Chillout.js is a standalone library, and you can use it in the most of JavaScript environments (Browser, Electron, Node.js etc.).

## Installation

### npm

```bash
$ npm install chillout
```

### CDN

chillout.js is available on [jsdelivr.com](https://www.jsdelivr.com/package/npm/chillout) and [cdnjs.com](https://cdnjs.com/libraries/chillout).

### Usage

```javascript
var chillout = require('chillout');

chillout.forEach([1, 2, 3], function(value) {
  console.log(value);
}).then(function() {
  console.log('done');
});

// 1
// 2
// 3
// 'done'
```

Object **chillout** is defined in the global scope if running in the browser window. ( `window.chillout` )

## Compatibility

The limiting factor for browser/node support is the use of `Promise`.  
You can use [es6-shim](https://github.com/paulmillr/es6-shim) or other `Promise` polyfills.

## async / await

You can write more simply using by `async/await` syntax.

Because all APIs in chillout.js return Promise, it is easy to handle with `async/await`.

## Benchmarks

Benchmarks the **ForStatement** and `chillout.repeat`.

```javascript
function heavyProcess() {
  var v;
  for (var i = 0; i < 5000; i++) {
    for (var j = 0; j < 5000; j++) {
      v = i * j;
    }
  }
  return v;
}
```

### ForStatement

```javascript
var time = Date.now();
for (var i = 0; i < 1000; i++) {
  heavyProcess();
}
var processingTime = Date.now() - time;
console.log(processingTime);
```

![CPU usage without chillout](https://raw.github.com/wiki/polygonplanet/chillout/images/benchmark-cpu-usage-without-chillout.png)

* Processing time: 107510ms.
* CPU usage on Node process (Average): **97.13%**

### chillout.repeat


```javascript
var time = Date.now();
chillout.repeat(1000, function(i) {
  heavyProcess();
}).then(function() {
  var processingTime = Date.now() - time;
  console.log(processingTime);
});
```

![CPU usage with chillout](https://raw.github.com/wiki/polygonplanet/chillout/images/benchmark-cpu-usage-using-chillout.png)

* Processing time: 138432ms.
* CPU usage on Node process (Average): **73.88%**


### Benchmark Result

![CPU usage with chillout](https://raw.github.com/wiki/polygonplanet/chillout/images/benchmark-cpu-usage-compare.png)

| &nbsp;                               | ForStatement | chillout.repeat |
| ------------------------------------ | ------------:| ---------------:|
| Processing time                      |    107510ms. |       138432ms. |
| CPU usage on Node process (Average)  |   **97.13%** |      **73.88%** |


You can confirm that `chillout.repeat` is running on a more low CPU usage than **ForStatement**.

chillout.js can run JavaScript in a natural speed with low CPU usage, but processing speed will be a bit slow.

One of the most important thing of performance in JavaScript, that is not numeric speed, but is to execute without causing stress to the user experience.


*(Benchmarks: chillout v3.1.2, Windows8.1 / Intel(R) Atom(TM) CPU Z3740 1.33GHz)*

### Run Benchmark

You can test benchmark with `npm run benchmark`.

----

## API

* [forEach](#foreach)
* [repeat](#repeat)
* [until](#until)
* [waitUntil](#waituntil)
* [forOf](#forof)

### forEach

Executes a provided function once per array or object element.  
The iteration will break if the callback function returns `chillout.StopIteration`, or an error occurs.  
This method can be called like JavaScript `Array forEach`.

* chillout.**forEach** ( obj, callback [, context ] )  
  * @param {_array|object_} _obj_ Target array or object  
  * @param {_function_} *callback* Function to execute for each element, taking three arguments:  
    - value: The current element being processed in the array/object
    - key: The key of the current element being processed in the array/object
    - obj: The array/object that `forEach` is being applied to
  * @param {_object_} [_context_] Value to use as `this` when executing callback  
  * @return {_promise_} Return new Promise

Example of array iteration:

```javascript
var values = ['a', 'b', 'c'];

chillout.forEach(values, function(value, key, obj) {
  console.log(value);
}).then(function() {
  console.log('done');
});

// 'a'
// 'b'
// 'c'
// 'done'
```


Example of object iteration:

```javascript
var values = {
  a: 1,
  b: 2,
  c: 3
};

chillout.forEach(values, function(value, key, obj) {
  console.log(key + ':' + value);
}).then(function() {
  console.log('done');
});

// 'a:1'
// 'b:2'
// 'c:3'
// 'done'
```

Example iteration using `async / await`:

Output all file contents, and finally output 'done'.


```javascript
async function getFileContents(url) {
  const response = await fetch(url);
  return response.text();
}

// Passing an async function as a callback in chillout.forEach
async function logFiles() {
  const files = ['/file1.txt', '/file2.txt', '/file3.txt'];
  await chillout.forEach(files, async url => {
    const contents = await getFileContents(url);
    console.log(contents);
  });
  console.log('done');
}

logFiles();
```


### repeat

Executes a provided function the specified number times.  
The iteration will break if the callback function returns `chillout.StopIteration`, or an error occurs.  
This method can be called like JavaScript `for` statement.

* chillout.**repeat** ( count, callback [, context ] )  
  * @param {_number|object_} _count_ The number of times or object for execute the function  
  Following parameters are available if specify object:
    - start: The number of start
    - step: The number of step
    - done: The number of done
  * @param {_function_} _callback_ Function to execute for each times, taking one argument:
    - i: The current number
  * @param {_object_} [_context_] Value to use as `this` when executing callback  
  * @return {_promise_} Return new Promise

Example of specify number:

```javascript
chillout.repeat(5, function(i) {
  console.log(i);
}).then(function() {
  console.log('done');
});

// 0
// 1
// 2
// 3
// 4
// 'done'
```

Example of specify object:

```javascript
chillout.repeat({ start: 10, step: 2, done: 20 }, function(i) {
  console.log(i);
}).then(function() {
  console.log('done');
});

// 10
// 12
// 14
// 16
// 18
// 'done'
```

Example iteration using `async / await`:

Output the user data from `/api/users/0` to `api/users/9`, and finally output 'done'.

```javascript
async function getUser(userId) {
  const response = await fetch(`/api/users/${userId}`);
  return response.json();
}

// Passing an async function as a callback in chillout.repeat
async function logUsers() {
  await chillout.repeat(10, async i => {
    const user = await getUser(i);
    console.log(user);
  });
  console.log('done');
}

logUsers();
```


### until

Executes a provided function until the `callback` returns `chillout.StopIteration`, or an error occurs.  
This method can be called like JavaScript `while (true) { ... }` statement.

* chillout.**until** ( callback [, context ] )  
  * @param {_function_} _callback_ The function that is executed for each iteration  
  * @param {_object_} [_context_] Value to use as `this` when executing callback  
  * @return {_promise_} Return new Promise

```javascript
var i = 0;
chillout.until(function() {
  console.log(i);
  i++;
  if (i === 5) {
    return chillout.StopIteration; // break loop
  }
}).then(function() {
  console.log('done');
});

// 0
// 1
// 2
// 3
// 4
// 'done'
```

Example iteration using `async / await`:

Watch for files changes, and finally output file changed contents.


```javascript
// Sleep until msec
function sleep(msec) {
  return new Promise(resolve => setTimeout(resolve, msec));
}

async function getFileContents(url) {
  const response = await fetch(url);
  return response.text();
}

// Passing an async function as a callback in chillout.until
async function logNewFileContents() {
  let previous = null;
  let contents = null;
  await chillout.until(async () => {
    contents = await getFileContents('./file1.txt');
    if (previous === null) {
      previous = contents;
    }
    if (contents !== previous) {
      console.log('file changed!');
      return chillout.StopIteration; // break loop
    }
    await sleep(1000);
  });
  console.log(contents);
  previous = contents;
}

logNewFileContents();
```

### waitUntil

Executes a provided function until the `callback` returns `chillout.StopIteration`, or an error occurs.  
This method can be called like JavaScript `while (true) { ... }` statement, and it works same as [`until`](#until), but it executes tasks with more slowly interval than `until` to reduce CPU load.  
This method is useful when you want to wait until some processing done.

* chillout.**waitUntil** ( callback [, context ] )  
  * @param {_function_} _callback_ The function that is executed for each iteration  
  * @param {_object_} [_context_] Value to use as `this` when executing callback  
  * @return {_promise_} Return new Promise

```javascript
chillout.waitUntil(function() {
  // Wait until the DOM body element is loaded
  if (document.body) {
    return chillout.StopIteration; // break loop
  }
}).then(function() {
  document.body.innerHTML += 'body loaded';
});
```


Example to wait until some processing is done.


```javascript
someProcessing();
chillout.waitUntil(function() {
  if (isSomeProcessingDone) {
    return chillout.StopIteration; // break loop
  }
}).then(function() {
  nextProcessing();
});
```

### forOf

Iterates the iterable objects, similar to the `for-of` statement.  
Executes a provided function once per element.  
The iteration will break if the callback function returns `chillout.StopIteration`, or an error occurs.

* chillout.**forOf** ( iterable, callback [, context ] )  
  * @param {_array|string|object_} _iterable_ Target iterable objects  
  * @param {_function_} _callback_ Function to execute for each element, taking one argument:
    - value: A value of a property on each iteration
  * @param {_object_} [_context_] Value to use as `this` when executing callback  
  * @return {_promise_} Return new Promise

Example of iterate array:

```javascript
chillout.forOf([1, 2, 3], function(value) {
  console.log(value);
}).then(function() {
  console.log('done');
});

// 1
// 2
// 3
// 'done'
```

Example of iterate string:

```javascript
chillout.forOf('abc', function(value) {
  console.log(value);
}).then(function() {
  console.log('done');
});

// 'a'
// 'b'
// 'c'
// 'done'
```

## Comparison Table

You can reduce the CPU load by using the chillout.js API instead native JavaScript loops in your application.

Examples:

| JavaScript Statement                 | chillout                                                                      |
| -------------------------------------|-------------------------------------------------------------------------------|
| [1, 2, 3].forEach(function(v, i) {}) | chillout.forEach([1, 2, 3], function(v, i) {})                                |
| for (i = 0; i < 5; i++) {}           | chillout.repeat(5, function(i) {})                                            |
| for (i = 10; i < 20; i += 2) {}      | chillout.repeat({ start: 10, step: 2, done: 20 }, function(i) {})              |
| while (true) {}                      | chillout.until(function() {})                                                  |
| while (cond()) {<br>&nbsp;&nbsp;doSomething();<br>}                    | chillout.until(function() {<br>&nbsp;&nbsp;if (!cond()) return chillout.StopIteration;<br>&nbsp;&nbsp;doSomething();<br>})    |
| for (value of [1, 2, 3]) {}          | chillout.forOf([1, 2, 3], function(value) {})                                 |

(You can write more simply using by `async / await` syntax because chillout.js' all APIs return Promise.)

## Contributing

We're waiting for your pull requests and issues.
Don't forget to execute `npm run test` before requesting.
Accepted only requests without errors.

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [5.0.0] - 2020-01-24
### Changed
- Changed delay time to run more slowly to reduce CPU than previously.

### Removed
- Drop bower support.

## [4.0.4] - 2019-11-02
### Fixed
- Bump dev dependencies.

## [4.0.3] - 2019-08-18
### Fixed
- Fixed invalid uglifyjs option.

## [4.0.2] - 2019-03-18
### Fixed
- Fixed tests timeout.

## [4.0.0] - 2019-03-03
### Added
- Added `waitUntil` method.
- Added core methods and properties to exports.

### Changed
- Changed `till` method name to `until`
- Changed the key name to `done` instead of `end` in `repeat` method with arguments passed as object.
- Changed use `return chillout.StopIteration` to stop the loop instead of `return false`.

### Fixed
- Fixed parameters can be passed when called with nested Promise.

## [3.1.9] - 2018-12-10
### Fixed
- Fixed `can't find module` using "require".

## [3.1.8] - 2018-12-10
### Fixed
- Fixed result when using `async/await`.

### Changed
- Change internal iteration stopper.

## [3.1.7] - 2018-05-10
### Fixed
- Fixed bug for `can't find module`.

## [3.1.6] - 2018-02-17
### Fixed
- Fixed build commands.

## [3.1.4] - 2017-07-15
### Changed
- Change eslint base to `eslint:recommended`.

## [3.1.3] - 2017-01-20
### Fixed
- Improve return statement in iterator function. Thanks to [@JuhQ](https://github.com/JuhQ).

## [3.1.0] - 2016-06-06
### Added
- Add forOf iterator.

## [3.0.0] - 2016-04-30
### Changed
- Change API method names.

## [2.0.0] - 2016-03-12
### Changed
- Modularize the code base

## [1.1.5] - 2016-02-12
### Fixed
- Fix chillout.each to work with generic array-like object instead of an array.

## [1.1.0] - 2016-01-15
### Fixed
- Fix setImmediate by MessageChannel.

## [1.0.0] - 2016-01-05
### Added
- First release.

[5.0.0]: https://github.com/polygonplanet/chillout/compare/4.0.3...5.0.0
[4.0.4]: https://github.com/polygonplanet/chillout/compare/4.0.3...4.0.4
[4.0.3]: https://github.com/polygonplanet/chillout/compare/4.0.2...4.0.3
[4.0.2]: https://github.com/polygonplanet/chillout/compare/4.0.0...4.0.2
[4.0.0]: https://github.com/polygonplanet/chillout/compare/3.1.9...4.0.0
[3.1.9]: https://github.com/polygonplanet/chillout/compare/3.1.8...3.1.9
[3.1.8]: https://github.com/polygonplanet/chillout/compare/3.1.7...3.1.8
[3.1.7]: https://github.com/polygonplanet/chillout/compare/3.1.6...3.1.7
[3.1.6]: https://github.com/polygonplanet/chillout/compare/3.1.4...3.1.6
[3.1.4]: https://github.com/polygonplanet/chillout/compare/3.1.3...3.1.4
[3.1.3]: https://github.com/polygonplanet/chillout/compare/3.1.0...3.1.3
[3.1.0]: https://github.com/polygonplanet/chillout/compare/3.0.0...3.1.0
[3.0.0]: https://github.com/polygonplanet/chillout/compare/2.0.0...3.0.0
[2.0.0]: https://github.com/polygonplanet/chillout/compare/1.1.5...2.0.0
[1.1.5]: https://github.com/polygonplanet/chillout/compare/1.1.0...1.1.5
[1.1.0]: https://github.com/polygonplanet/chillout/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/polygonplanet/chillout/releases/tag/1.0.0


chillout.js
/**
 * chillout
 *
 * @description Reduce JavaScript CPU usage by asynchronous iteration
 * @version     1.1.5
 * @date        2016-02-12
 * @link        https://github.com/polygonplanet/chillout
 * @copyright   Copyright (c) 2016 polygon planet <polygon.planet.aqua@gmail.com>
 * @license     MIT
 */

/*jshint bitwise:false, eqnull:true */
(function(name, context, factory) {

  // Supports AMD, Node.js, CommonJS and browser context
  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      module.exports = factory();
    } else {
      exports[name] = factory();
    }
  } else if (typeof define === 'function' && define.amd) {
    define(factory);
  } else {
    context[name] = factory();
  }

}('chillout', this, function() {
  'use strict';

  var chillout = {};

  // nextTick implementation
  if (typeof setImmediate === 'function') {
    chillout.nextTick = function(task) {
      setImmediate(task);
    };
  } else if (typeof process === 'object' && typeof process.nextTick === 'function') {
    chillout.nextTick = function(task) {
      process.nextTick(task);
    };
  } else if (typeof MessageChannel === 'function') {
    chillout.nextTick = (function() {
      // http://www.nonblocking.io/2011/06/windownexttick.html
      var channel = new MessageChannel();
      var head = {}, tail = head;

      channel.port1.onmessage = function() {
        head = head.next;
        var task = head.task;
        delete head.task;
        task();
      };

      return function(task) {
        tail = tail.next = {task: task};
        channel.port2.postMessage(0);
      };
    }());
  } else {
    chillout.nextTick = function(task) {
      setTimeout(task, 0);
    };
  }


  var iterator = {
    each: function(obj, callback, context) {
      var i = 0;
      var len;

      if (isArrayLike(obj)) {
        len = obj.length;

        return {
          next: function() {
            if (i >= len) {
              return false;
            }
            return callback.call(context, obj[i], i++, obj);
          }
        };
      }

      var keys = Object.keys(obj);
      len = keys.length;

      return {
        next: function() {
          if (i >= len) {
            return false;
          }

          var key = keys[i++];
          return callback.call(context, obj[key], key, obj);
        }
      };
    },
    repeat: function(count, callback, context) {
      var i, step, end;

      if (count && typeof count === 'object') {
        i = count.start || 0;
        step = count.step || 1;
        end = count.end;
      } else {
        i = 0;
        step = 1;
        end = count;
      }

      return {
        next: function() {
          var res = callback.call(context, i);

          i += step;
          if (i >= end) {
            return false;
          }
          return res;
        }
      };
    },
    forever: function(callback, context) {
      return {
        next: function() {
          return callback.call(context);
        }
      };
    }
  };


  function iterate(it) {
    return new Promise(function(resolve, reject) {
      var totalTime = 0;

      function then(value) {
        value.then(function(res) {
          if (res === false) {
            resolve();
          } else {
            _iterate();
          }
        }, function(err) {
          reject(err);
        });
      }

      function _iterate() {
        var cycleStartTime = Date.now();
        var res, risk, margin, delay, time, endTime, cycleEndTime;

        try {
          for (;;) {
            res = it.next();
            if (res === false) {
              resolve();
              return;
            }

            if (isThenable(res)) {
              then(res);
              return;
            }

            endTime = Date.now();
            cycleEndTime = endTime - cycleStartTime;
            totalTime += cycleEndTime;

            if (totalTime > 1000) {
              // Break the loop when the process is continued for more than 1s
              break;
            }

            if (cycleEndTime < 10) {
              // Delay is not required for fast iteration
              continue;
            }

            risk = Math.min(10, Math.floor(cycleEndTime / 10));
            margin = endTime % (10 - risk);
            if (!margin) {
              // Break the loop if processing has exceeded the allowable
              break;
            }
          }
        } catch (e) {
          reject(e);
          return;
        }

        // Add delay corresponding to the processing speed
        time = Math.sqrt(cycleEndTime) * Math.min(1000, cycleEndTime) / 80;
        delay = Math.min(1000, Math.floor(time));
        totalTime = 0;

        if (delay > 10) {
          setTimeout(_iterate, delay);
        } else {
          chillout.nextTick(_iterate);
        }
      }

      chillout.nextTick(_iterate);
    });
  }


  /**
   * Executes a provided function once per array or object element.
   * The iteration will break if the callback function returns `false`.
   *
   * @param {Array|Object} obj Target array or object.
   * @param {Function} callback Function to execute for each element,
   *   taking three arguments:
   * - value: The current element being processed in the array/object.
   * - key: The key of the current element being processed in the array/object.
   * - obj: The array/object that `each` is being applied to.
   * @param {Object} [context] Value to use as `this` when executing callback.
   * @return {Promise} Return new Promise.
   */
  chillout.each = function(obj, callback, context) {
    return iterate(iterator.each(obj, callback, context));
  };

  /**
   * Executes a provided function the specified number times.
   * The iteration will break if the callback function returns `false`.
   *
   * @param {number|Object} count The number of times or object for execute
   *   the function. Following parameters are available if specify object:
   * - start: The number of start.
   * - step: The number of step.
   * - end: The number of end.
   * @param {Function} callback Function to execute for each times,
   *   taking an argument:
   * - i: The current number.
   * @param {Object} [context] Value to use as `this` when executing callback.
   * @return {Promise} Return new Promise.
   */
  chillout.repeat = function(count, callback, context) {
    return iterate(iterator.repeat(count, callback, context));
  };

  /**
   * Executes a provided function forever.
   * The iteration will break if the callback function returns `false`.
   *
   * @param {Function} callback The function that is executed for each
   *   iteration.
   * @param {Object} [context] Value to use as `this` when executing callback.
   * @return {Promise} Return new Promise.
   */
  chillout.forever = function(callback, context) {
    return iterate(iterator.forever(callback, context));
  };


  function isThenable(x) {
    return x != null && typeof x.then === 'function';
  }

  function isArrayLike(x) {
    return x != null && typeof x.length === 'number' && x.length >= 0;
  }

  return chillout;
}));
